### PR TITLE
Only allow one addr hash per challenge interval blocks

### PR DESCRIPTION
**Pull request by Vagabond**

### CODE REVIEW

File: src/poc/miner_poc_statem.erl [hunk 1]
```diff
@@ -75,10 +75,12 @@
     retry = ?CHALLENGE_RETRY :: non_neg_integer(),
     receipts_timeout = ?RECEIPTS_TIMEOUT :: non_neg_integer(),
     poc_restarts = ?POC_RESTARTS :: non_neg_integer(),
-    txn_ref = make_ref() :: reference()
+    txn_ref = make_ref() :: reference(),
+    addr_hash_filter :: undefined | {pos_integer(), pos_integer(), pos_integer(), binary(), bloom_nif:bloom()}
```
Comment (evanmcc) [inline src/poc/miner_poc_statem.erl:79]: can we do a module local record for this instead of a tuple?

File: src/poc/miner_poc_statem.erl [hunk 2]
```diff
@@ -75,10 +75,12 @@
...
 }).
 
--type state() :: requesting | mining | receiving | waiting.
+-
+type state() :: requesting | mining | receiving | waiting.
```
Comment (evanmcc) [inline src/poc/miner_poc_statem.erl:83]: typo?

File: src/poc/miner_poc_statem.erl [hunk 3]
```diff
@@ -610,6 +634,159 @@ load_data(BaseDir) ->
...
+    lager:info("no filter"),
+    undefined;
+check_addr_hash(PeerAddr, #data{addr_hash_filter={Start, Height, Size, Hash, Bloom}}) ->
+    case multiaddr:protocols(PeerAddr) of
+        [{"ip4",Address},{_,_}] ->
+            {ok, Addr} = inet:parse_ipv4_address(Address),
+            Val = binary:part(enacl:pwhash(list_to_binary(tuple_to_list(Addr)), binary:part(Hash, {0, enacl:pwhash_SALTBYTES()})), {0, Size}),
+            case bloom:check_and_set(Bloom, Val) of
+                true ->
+                    lager:info("~p filtering peer ~p ~p", [{Start, Height}, PeerAddr, Val]),
```
Comment (evanmcc) [inline src/poc/miner_poc_statem.erl:779]: debugging

File: src/poc/miner_poc_statem.erl [hunk 4]
```diff
@@ -610,6 +634,159 @@ load_data(BaseDir) ->
...
+                    {ok, CurHeight} = blockchain:height(Blockchain),
+                    case max(Height - (Height rem Interval), 1) of
+                        StartHeight ->
+                            case CurHeight of
+                                Height ->
+                                    %% ok, everything lines up
+                                    Data;
+                                _ ->
+                                    case blockchain:get_block(Height, Blockchain) of
+                                        {ok, Block} ->
```
Comment (evanmcc) [inline src/poc/miner_poc_statem.erl:705]: much sprawl wow

File: src/poc/miner_poc_statem.erl [hunk 5]
```diff
@@ -610,6 +634,159 @@ load_data(BaseDir) ->
...
+                    {ok, Height} = blockchain:height(Blockchain),
+                    StartHeight = max(Height - (Height rem Interval), 1),
+                    %% check if we have this block
+                    case blockchain:get_block(StartHeight, Blockchain) of
+                        {ok, Block} ->
+                            Hash = blockchain_block:hash_block(Block),
+                            %% ok, now we can build the filter
+                            Gateways = blockchain_ledger_v1:gateway_count(Ledger),
+                            {ok, Bloom} = bloom:new_optimal(Gateways, 1.0e-6),
+                            blockchain:fold_chain(fun(Blk, _) ->
```
Comment (evanmcc) [inline src/poc/miner_poc_statem.erl:658]: can we dedupe this init stuff?

File: src/poc/miner_poc_statem.erl [hunk 6]
```diff
@@ -252,20 +254,20 @@ receiving(info, {blockchain_event, {add_block, _Hash, _, _}}, #data{receiving_ti
...
             catch C:E:S ->
                     lager:warning("error ~p:~p submitting requests: ~p", [C, E, S]),
-                    {next_state, requesting, save_data(Data#data{state=requesting})}
+                    {next_state, requesting, save_data(maybe_init_addr_hash(Data#data{state=requesting}))}
             end
     end;
 receiving(info, {blockchain_event, {add_block, _Hash, _, _}}, #data{receiving_timeout=T}=Data) ->
     lager:info("got block ~p decreasing timeout", [_Hash]),
-    {keep_state, save_data(Data#data{receiving_timeout=T-1})};
+    {keep_state, save_data(maybe_init_addr_hash(Data#data{receiving_timeout=T-1}))};
```
Comment (evanmcc) [inline src/poc/miner_poc_statem.erl:270]: I feel like we should be able to do this just in load and init?

**Review by evanmcc — COMMENTED** [review]
this looks all right needs some cleanup and I need another look when I'm fresher.

File: src/poc/miner_poc_statem.erl (hunk 6 above)
Comment (Vagabond) [reply src/poc/miner_poc_statem.erl:270]: No as the filter might need to be re-intited on a block height change.

File: src/poc/miner_poc_statem.erl (hunk 5 above)
Comment (Vagabond) [reply src/poc/miner_poc_statem.erl:658]: yeah, I'll take care of it

File: src/poc/miner_poc_statem.erl (hunk 1 above)
Comment (Vagabond) [reply src/poc/miner_poc_statem.erl:79]: Yep

File: src/poc/miner_onion_handler.erl [hunk 7]
```diff
@@ -28,7 +28,7 @@
 
 -include("miner_util.hrl").
 
--record(state, {peer :: undefined | libp2p_crypto:pubkey_bin()}).
+-record(state, {peer :: undefined | libp2p_crypto:pubkey_bin(), peer_addr}).
```
Comment (vihu) [inline src/poc/miner_onion_handler.erl:31]:
```suggestion
-record(state,
    { peer :: undefined | libp2p_crypto:pubkey_bin(),
      peer_addr :: undefined | string()
     }).
```

File: src/poc/miner_poc_handler.erl [hunk 8]
```diff
@@ -28,7 +28,7 @@
 
 -include("miner_util.hrl").
 
--record(state, {peer :: undefined | libp2p_crypto:pubkey_bin()}).
+-record(state, {peer :: undefined | libp2p_crypto:pubkey_bin(), peer_addr}).
```
Comment (vihu) [inline src/poc/miner_poc_handler.erl:31]:
```suggestion
-record(state,
    { peer :: undefined | libp2p_crypto:pubkey_bin(),
      peer_addr :: undefined | string()
     }).
```

File: src/poc/miner_poc_statem.erl [hunk 9]
```diff
@@ -610,6 +634,159 @@ load_data(BaseDir) ->
...
+                {ok, Interval} ->
+                    {ok, Height} = blockchain:height(Blockchain),
+                    StartHeight = max(Height - (Height rem Interval), 1),
+                    %% check if we have this block
+                    case blockchain:get_block(StartHeight, Blockchain) of
+                        {ok, Block} ->
+                            Hash = blockchain_block:hash_block(Block),
+                            %% ok, now we can build the filter
+                            Gateways = blockchain_ledger_v1:gateway_count(Ledger),
+                            {ok, Bloom} = bloom:new_optimal(Gateways, 1.0e-6),
```
Comment (vihu) [inline src/poc/miner_poc_statem.erl:657]: Can we hoist `1.0e-6` up the module as a constant value?

File: src/poc/miner_poc_statem.erl [hunk 10]
```diff
@@ -610,6 +634,159 @@ load_data(BaseDir) ->
...
+                _ ->
+                    lager:info("no challenge interval"),
+                    Data
+            end;
+        _ ->
+            lager:info("no hash bytes size"),
+            Data#data{addr_hash_filter=undefined}
+    end.
+
+update_addr_hash(Bloom, Element) ->
```
Comment (vihu) [inline src/poc/miner_poc_statem.erl:745]:
```suggestion
-spec update_addr_hash(Bloom :: bloom_nif:bloom(), Element :: blockchain_poc_path_element_v1:poc_element()) -> ok.
update_addr_hash(Bloom, Element) ->
```

File: src/poc/miner_poc_statem.erl [hunk 11]
```diff
@@ -610,6 +634,159 @@ load_data(BaseDir) ->
...
+                                                                                                                 false
+                                                                                                         end),
+                                                                          case Blk == Block of
+                                                                              true ->
+                                                                                  return;
+                                                                              false ->
+                                                                                  continue
+                                                                          end
+                                                                  end, any, element(2, blockchain:head_block(Blockchain)), Blockchain),
+                                            lager:info("updated filter"),
```
Comment (vihu) [inline src/poc/miner_poc_statem.erl:724]: debugging..

**Review by evanmcc — APPROVED** [review]

**Pull request merged**